File: leetcode-js/306-additive-number.js
/**
 * @param {string} num
 * @return {boolean}
 */
var isAdditiveNumber = function(num) {
    for (let i = 1; i <= Math.floor(num.length / 2); i += 1) {
        for (let j = 1; j <= Math.floor((num.length - 1) / 2); j += 1) {
            if (check(num.substr(0, i), num.substr(i, j), num.substr(i + j))) {
                return true;
            }
        }
    }
    return false;
};

function check(num1, num2, num) {
    if (num1.length > 1 && num1[0] === '0' || num2.length > 1 && num2[0] === '0') {
        return false;
    }
    const sum = add(num1, num2);
    if (num === sum) {
        return true;
    }
    if (num.length <= sum.length || sum !== num.substr(0, sum.length)) {
        return false;
    } else {
        return check(num2, sum, num.substr(sum.length));
    }
}

function add(n, m) {
    return (parseInt(n) + parseInt(m)).toString();
}

console.log(isAdditiveNumber("112358"));
console.log(isAdditiveNumber("101"));

File: leetcode-java/306-additive-number.java
public class Solution {
    public boolean isAdditiveNumber(String num) {
        int n = num.length();

        for (int i = 1; i <= n / 2; i++) {
            for (int j = 1; Math.max(i, j) <= n - i - j; j++) {
                if (isValid(num, i, j)) {
                    return true;
                }
            }
        }

        return false;
    }

    boolean isValid(String num, int len1, int len2) {
        if (num.charAt(0) == '0' && len1 > 1) {
            return false;
        }
        if (num.charAt(len1) == '0' && len2 > 1) {
            return false;
        }

        Long x1 = Long.parseLong(num.substring(0, len1));
        Long x2 = Long.parseLong(num.substring(len1, len1 + len2));
        int pos = len1 + len2;

        while (pos < num.length()) {
            x2 = x1 + x2;
            x1 = x2 - x1;
            String sum = x2.toString();
            if (!num.startsWith(sum, pos)) {
                return false;
            }
            pos += sum.length();
        }

        return true;
    }
}
